libraries/AP_Filesystem/AP_Filesystem_FlashMemory_LittleFS.cpp

```diff
@@ -0,0 +1,1227 @@
+        return false;
+    }
+
+    WITH_SEMAPHORE(dev_sem);
+
+    // Read manufacturer ID
+    uint8_t cmd = JEDEC_DEVICE_ID;
+    uint8_t buf[4];
+    dev->transfer(&cmd, 1, buf, 4);
+
+#if AP_FILESYSTEM_LITTLEFS_FLASH_TYPE == AP_FILESYSTEM_FLASH_W25NXX
+    uint32_t id = buf[1] << 16 | buf[2] << 8 | buf[3];
+
+#else
+    uint32_t id = buf[0] << 16 | buf[1] << 8 | buf[2];
+#endif
+
+    // Let's specify the terminology here.
+    //
+    // 1 block = smallest unit that we can _erase_ in a single operation
+    // 1 page = smallest unit that we can read or program in a single operation
+    //
+    // So, for instance, if we have 4K sectors on the flash chip and we can
+    // always erase a single 4K sector, the LFS block size will be 4096 bytes,
+
+    // irrespectively of what the flash chip documentation refers to as a "block"
+    /* Most flash chips are programmable in chunks of 256 bytes and erasable in
+     * blocks of 4K so we start with these defaults */
+    uint16_t page_size = 256;
+    flash_block_size = 4096;
+    flash_block_count = 0;
+
+    switch (id) {
+#if AP_FILESYSTEM_LITTLEFS_FLASH_TYPE == AP_FILESYSTEM_FLASH_W25NXX
+    case JEDEC_ID_WINBOND_W25N01GV:
+        /* 128M, programmable in chunks of 2048 bytes, erasable in blocks of 128K */
+        page_size = 2048;
+        flash_block_size = 131072;
+        flash_block_count = 1024;
+        break;
+    case JEDEC_ID_WINBOND_W25N02KV:
+        /* 256M, programmable in chunks of 2048 bytes, erasable in blocks of 128K */
+        page_size = 2048;
+        flash_block_size = 131072;
+        flash_block_count = 2048;
+        break;
+#else
+    case JEDEC_ID_WINBOND_W25Q16:
+    case JEDEC_ID_MICRON_M25P16:
+        flash_block_count = 32;   /* 128K */
+        break;
+
+    case JEDEC_ID_WINBOND_W25Q32:
+    case JEDEC_ID_WINBOND_W25X32:
+    case JEDEC_ID_MACRONIX_MX25L3206E:
+        flash_block_count = 64;   /* 256K */
+        break;
+
+    case JEDEC_ID_MICRON_N25Q064:
+    case JEDEC_ID_WINBOND_W25Q64:
+    case JEDEC_ID_MACRONIX_MX25L6406E:
+        flash_block_count = 128;  /* 512K */
+        break;
+
+    case JEDEC_ID_MICRON_N25Q128:
+    case JEDEC_ID_WINBOND_W25Q128:
+    case JEDEC_ID_WINBOND_W25Q128_2:
+    case JEDEC_ID_CYPRESS_S25FL128L:
+        flash_block_count = 256;  /* 1M */
+        break;
+
+    case JEDEC_ID_WINBOND_W25Q256:
+    case JEDEC_ID_MACRONIX_MX25L25635E:
+        flash_block_count = 512;  /* 2M */
+        use_32bit_address = true;
+        break;
+#endif
+    default:
+        hal.scheduler->delay(2000);
+        printf("Unknown SPI Flash 0x%08x\n", id);
+        return 0;
+    }
+
+    fs_cfg.read_size = page_size;
+    fs_cfg.prog_size = page_size;
+    fs_cfg.block_size = flash_block_size * LFS_FLASH_BLOCKS_PER_BLOCK;
+    fs_cfg.block_count = flash_block_count / LFS_FLASH_BLOCKS_PER_BLOCK;
+#if AP_FILESYSTEM_LITTLEFS_FLASH_TYPE == AP_FILESYSTEM_FLASH_W25NXX
+    fs_cfg.metadata_max = page_size;
```

Comment:
page size is 2k. I tried 8k before and it was slower because of the extra page copies when un-inlining. One option would be to specify unlined for logs. It exposes more flags at the logging layer but might be worth it.

Reply:
Could also reduce the inline file size.